src/groupeddataframe/grouping.jl

```diff
@@ -245,39 +337,110 @@ end
 
 combine(gd::GroupedDataFrame) = combine(identity, gd)
 
-function _combine(first::Union{NamedTuple, DataFrameRow}, f::Function,
+function _combine(f::Union{AbstractVector{<:Pair}, Tuple{Vararg{Pair}},
+                           NamedTuple{<:Any, <:Tuple{Vararg{Pair}}}},
                   gd::GroupedDataFrame)
-    m = length(first)
-    n = length(gd)
-    idx = Vector{Int}(undef, n)
-    initialcols = ntuple(i -> Tables.allocatecolumn(typeof(first[i]), n), m)
-    cols = _combine!(first, initialcols, idx, 1, 1, f, gd, tuple(propertynames(first)...))
-    valscat = DataFrame(collect(cols), collect(propertynames(first)))
+    idxs = (first(p) isa Union{Integer, Symbol} ?
+            index(gd.parent)[first(p)] :
+            index(gd.parent)[collect(first(p))] for p in f)
+    incols = [names(gd.parent)[idx] for idx in idxs]
+    allcols = collect(reduce(union, (nam isa Symbol ? (nam,) : nam for nam in incols)))
+    f2 = ntuple(length(f)) do i
+        nms = incols[i]
+        (nms isa AbstractArray ? Tuple(nms) : nms) => last(f[i])
+    end
+    # Use temporary names for columns, and rename after the fact where appropriate
+    tmpnms = Tuple(gennames(length(f)))::NTuple{length(f), Symbol}
+    fun = function(x)
+        tup = map(f2) do p
+            local incols = first(p)
+            res = incols isa Tuple ?
+                  last(p)(NamedTuple{incols}(map(c -> x[c], incols))) :
+                  last(p)(x[incols])
+            if res isa Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractVector, AbstractMatrix}
+                throw(ArgumentError("a single value result is required when passing a vector or tuple " *
```

Comment:
Why we do not allow returning `AbstractVector`?

I can agree that we do not allow other forms (and this answers my question above).

Reply:
I see this as an important problematic case - why don't you want to accept returning an `AbstractVector`?

Reply:
Just to make my life simpler. :-)

But now that everything has settled, I've added that feature.